test/ejabberd_tests/tests/connect_SUITE.erl

```diff
@@ -338,3 +355,99 @@ default_context(To) ->
 node2_rpccall(Module, Function, Args) ->
     Node = ct:get_config(ejabberd2_node),
     rpc:call(Node, Module, Function, Args).
+
+%% Tests featuress advertisement
+stream_features_test(Config) ->
+    UserSpec = escalus_users:get_userspec(Config, ?SECURE_USER),
+    List = [start_stream, stream_features, {?MODULE, verify_features}],
+    escalus_connection:start(UserSpec, List),
+    ok.
+
+verify_features(Conn, Props, Features) ->
+    %% should not advertise compression before tls
+    ?assert_equal(false, has_feature(compression, Features)),
+    %% start tls. Starttls should be then removed from list and compression should be added
+    {Conn1, Props1} = escalus_session:starttls(Conn, Props),
+    {Conn2, Props2, Features2} = escalus_session:stream_features(Conn1, Props1, []),
+    ?assert_equal(false, has_feature(starttls, Features2)),
+    ?assert(false =/= has_feature(compression, Features2)),
+    %% start compression. Compression should be then removed from list
+    {Conn3, Props3} = escalus_session:compress(Conn2, Props2),
+    {Conn4, Props4, Features4} = escalus_session:stream_features(Conn3, Props3, []),
+    ?assert_equal(false, has_feature(compression, Features4)),
+    ?assert_equal(false, has_feature(starttls, Features4)),
+    %%
+    escalus_session:authenticate(Conn4, Props4, Features4).
+
+has_feature(Feature, FeatureList) ->
+    {_, Value} = lists:keyfind(Feature, 1, FeatureList),
+    Value.
+
+%% tests for different order feature negotiation
+tls_compression_authenticate(Config) ->
```

Comment:
According to XEP-0138 we shouldn't allow compression before authentication. It may be used by malicious and not-registered users to attack the server.